docs/rfc/0000-cli-infrastructure.adoc

```diff
@@ -0,0 +1,254 @@
+implemented:
+
+. `ssh-agent` compatability with `<<radicle-keystore>>`, which has been
+started in `<<rk-17, radicle-keystore#17>>`.
+. A subcommand `rad-key-ssh-add` for adding one's secret key to the
+`ssh-agent` using the supplied constraints, e.g. ask for password.
+. `radicle-link#461` to have a read-only storage.
+
+== Library/Command Organisation
+
+As proposed here, `radicle-link` will provide a core set of
+functionality. Something this document wishes to standardise is the
+organisation of these components. The author sees two options:
+. One binary that defines the core subcommands
+. Many binary that take advantage of the subcommand mechanism
+
+Pros of *1.*:
+
+* One-stop shop for all core functionality
+* One binary is packages and distributed
+
+Cons of *1.*:
+
+* Making changes to one subset of the commands could mean a version
+  change of the whole binary
+
+Pros of *2.*:
+
+* We will be organising commands as libraries (as per RFC-682) — so
+  modular commands would mirror this structure well.
+* Subcommands can be versioned independently of each other.
+* We can dogfood our subcommand functionality
+
+Cons of *2.*:
+
+* Makes it easier to override one of the core commands.
+* Many binary must be packaged and distributed.
+
+The author recommends *2.*. The modularity provides a clean separation
+of concerns and avoids a monolothic approach. We assume good faith
+of overrides not happening [TODO: is there some way we can just
+disallow this programatically?]. For packaging and distribution of the
+binaries, we propose to have a script that will install each
+subcommand binary — until we have a more well-founded way of
+distributing Radicle artifacts.
+
+=== `radicle-bins`
+
+Per RFC-682#CLI, all subcommands will be provided as
+libraries. However, this leaves the question of where the binary might
+live. So far, it has been an unspoken formalism that all binaries live
+in `radicle-bins` where we can provide a `Cargo.lock` file. All
+functionality will live in `radicle-link`, including the `main`
+function, and the entry in `radicle-bins` will be a thin wrapper
+around this function.
+
+=== Distribution
+
+We are noting here that -- at the time of writing -- we cannot publish
+these binaries to <<crates.io>> due to the fact that `librad` and other
+crucial libraries only live as a `git` repositories and are not on
+crates.io. Anything that depends on them cannot be published as a
+consequence.
+
+This means that all subcommands will need to be built from source
+initially, unless we build the binaries and distribute them from a
+well-known source.
+
+== Core
+
+With all the details in place, we define a non-exhaustive list of the
+core commands that we wish to include for the first iteration. More
+components MAY be added as the result of future RFCs. We SHALL keep a
+list of the core commands as part of the `radicle-link` documentation
+— as part of the `README`.
+
+=== `rad-project`
```

Comment:
I don't think project/person as separate subcommands is a smart idea, because that will just repeat most of the commonalities. It also prevents extensibility without having to change the CLI for the extension specifically. 

A more maintainable approach would be to either disambiguate with a `--type` argument, or add another layer of sub-command-ness (eg. `rad-identities project [create|...]`). Or else, move CRUD-style verbs to the top-level, like `rad create project [opts]`.

Reply:
> It also prevents extensibility without having to change the CLI for the extension specifically.

Can you elaborate on what you mean here?

The rest makes sense, but I'm still mulling over what the consequences of one option over the other might be :) I think I'll need to sketch them out to see which makes more sense

Reply:
I mean that extension payloads are basically opaque JSON blobs. One may also want to pipe an entire payload as JSON, and have the CLI infer the type. Or maybe some new identity kind was introduced, but the local CLI installation doesn’t know about it yet (disregard for a moment that the replication logic needs to know about it — maybe that namespace was obtained by some other means). We could still make it so the CLI is able to display _something_.

Reply:
Ah that's clear, thanks for the insight!